zetaclient/chains/ton/observer/inbound.go

```diff
@@ -0,0 +1,261 @@
+package observer
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"cosmossdk.io/math"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/tonkeeper/tongo/ton"
+
+	"github.com/zeta-chain/node/pkg/coin"
+	toncontracts "github.com/zeta-chain/node/pkg/contracts/ton"
+	"github.com/zeta-chain/node/pkg/ticker"
+	"github.com/zeta-chain/node/zetaclient/chains/ton/liteapi"
+	zctx "github.com/zeta-chain/node/zetaclient/context"
+	"github.com/zeta-chain/node/zetaclient/zetacore"
+)
+
+const (
+	// MaxTransactionsPerTick is the maximum number of transactions to process on a ticker
+	MaxTransactionsPerTick = 100
+)
+
+func (ob *Observer) watchInbound(ctx context.Context) error {
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	var (
+		chainID         = ob.Chain().ChainId
+		initialInterval = ticker.SecondsFromUint64(ob.ChainParams().InboundTicker)
+		sampledLogger   = ob.Logger().Inbound.Sample(&zerolog.BasicSampler{N: 10})
+	)
+
+	ob.Logger().Inbound.Info().Msgf("WatchInbound started for chain %d", chainID)
+
+	task := func(ctx context.Context, t *ticker.Ticker) error {
+		if !app.IsInboundObservationEnabled() {
+			sampledLogger.Info().Msgf("WatchInbound: inbound observation is disabled for chain %d", chainID)
+			return nil
+		}
+
+		if err := ob.observeInbound(ctx); err != nil {
+			ob.Logger().Inbound.Err(err).Msg("WatchInbound: observeInbound error")
+		}
+
+		newInterval := ticker.SecondsFromUint64(ob.ChainParams().InboundTicker)
+		t.SetInterval(newInterval)
+
+		return nil
+	}
+
+	return ticker.Run(
+		ctx,
+		initialInterval,
+		task,
+		ticker.WithStopChan(ob.StopChannel()),
+		ticker.WithLogger(ob.Logger().Inbound, "WatchInbound"),
+	)
+}
+
+func (ob *Observer) observeInbound(ctx context.Context) error {
+	if err := ob.ensureLastScannedTX(ctx); err != nil {
+		return errors.Wrap(err, "unable to ensure last scanned tx")
+	}
+
+	// extract logicalTime and tx hash from last scanned tx
+	lt, hashBits, err := liteapi.TransactionHashFromString(ob.LastTxScanned())
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse last scanned tx %q", ob.LastTxScanned())
+	}
+
+	txs, err := ob.client.GetTransactionsSince(ctx, ob.gateway.AccountID(), lt, hashBits)
```

Comment:
wanted to know, are `txs` returned by `GetTransactionsSince` having finalized commitment level in state (or having enough confirmations)?

Reply:
There's no concept of confirmations in TON nor chain reogs. If tx gets into the block, it's safe to assume that it was finalized. 

The main trick here is that your tx might send messages to other contracts that might invoke logic and also send messages ... resulting in multiple txs across **several** blocks (think of ~~Uniswap~~ STONfi DEX swap). However, our gateway is simple and works by receiving a single internal message w/o any other actions required.

Reply:
got it